### PR TITLE
Optimization: allocate immutable Expression objects statically; ValueFieldAccessorResolver

### PR DESCRIPTION
**Pull request by SergeiPavlov**
* Reuse Expression objects, they are immutable.
* Use dictionary in `ValueFieldAccessorResolver` instead of linear search.

### CODE REVIEW

File: Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs [hunk 1]
```diff
@@ -50,48 +51,43 @@ private static class ValueFieldAccessorResolver
...
-          ReferenceEquals(type, WellKnownTypes.NullableDateTime) ? DateTimeAccessor :
-          ReferenceEquals(type, WellKnownTypes.NullableTimeSpan) ? TimeSpanAccessor :
-          ReferenceEquals(type, WellKnownTypes.NullableDecimal) ? DecimalAccessor :
-          ReferenceEquals(type, WellKnownTypes.NullableGuid) ? GuidAccessor : null;
-
-        return (probeType.MetadataToken ^ NullableTypeMetadataToken) == 0
-          ? ResolveByNullableType(probeType)
-          : ResolveByType(probeType);
-      }
+      private static readonly IReadOnlyDictionary<Type, ValueFieldAccessor> accessorByType =
```
Comment (botinko) [inline Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs:54]: If this place is really performance-critical, then if-else approach is better. I've just made bench

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 9 5900HX with Radeon Graphics, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.100-rc.1.21458.32
  [Host]     : .NET 5.0.10 (5.0.1021.41214), X64 RyuJIT
  DefaultJob : .NET 5.0.10 (5.0.1021.41214), X64 RyuJIT

|   Method |      type |      Mean |     Error |    StdDev | Code Size | Allocated |
|--------- |---------- |----------:|----------:|----------:|----------:|----------:|
| BenchOld |   Boolean |  2.883 ns | 0.0326 ns | 0.0289 ns |      63 B |         - |
| BenchNew |   Boolean | 11.759 ns | 0.2015 ns | 0.1683 ns |     211 B |         - |
| BenchOld |     Int16 |  3.286 ns | 0.0412 ns | 0.0385 ns |      63 B |         - |
| BenchNew |     Int16 | 12.485 ns | 0.2668 ns | 0.2365 ns |     211 B |         - |
| BenchOld |    SByte? |  3.602 ns | 0.1257 ns | 0.1634 ns |      63 B |         - |
| BenchNew |    SByte? | 11.759 ns | 0.1471 ns | 0.1304 ns |     211 B |         - |
| BenchOld | TimeSpan? |  4.364 ns | 0.0454 ns | 0.0402 ns |      63 B |         - |
| BenchNew | TimeSpan? | 11.150 ns | 0.2586 ns | 0.2419 ns |     211 B |         - |
| BenchOld |   UInt64? |  4.029 ns | 0.0759 ns | 0.0633 ns |      63 B |         - |
| BenchNew |   UInt64? | 11.801 ns | 0.2760 ns | 0.3389 ns |     211 B |         - |
| BenchOld |  TimeSpan |  4.328 ns | 0.0922 ns | 0.0863 ns |      63 B |         - |
| BenchNew |  TimeSpan | 12.453 ns | 0.1657 ns | 0.1384 ns |     211 B |         - |
```

Comment (botinko) [reply Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs:54]: even for `TimeSpan?` which is almost latest in if-else list, result is better. 
looks like if-else approach is be better for up to ~50 items.

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs [hunk 2]
```diff
@@ -20,6 +21,12 @@ namespace Xtensive.Orm.Rse.Transformation
 {
   internal sealed class RedundantColumnRemover : ColumnMappingInspector
   {
+    private static readonly MethodInfo selectMethodInfo = WellKnownTypes.Enumerable
+      .GetMethods()
+      .Single(methodInfo => methodInfo.Name == nameof(Enumerable.Select)
+        && methodInfo.GetParameters()[1].ParameterType.GetGenericTypeDefinition() == typeof(Func<,>))
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs:27]: It looks like we should move the `typeof(Func<,>)` to well-known types or at least to a static field.

File: Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs [hunk 3]
```diff
@@ -50,48 +51,43 @@ private static class ValueFieldAccessorResolver
...
+          [WellKnownTypes.NullableUInt64] = UInt64Accessor,
+          [WellKnownTypes.NullableSingle] = SingleAccessor,
+          [WellKnownTypes.NullableDouble] = DoubleAccessor,
+          [WellKnownTypes.NullableDateTime] = DateTimeAccessor,
+          [WellKnownTypes.NullableTimeSpan] = TimeSpanAccessor,
+          [WellKnownTypes.NullableDecimal] = DecimalAccessor,
+          [WellKnownTypes.NullableGuid] = GuidAccessor,
+        };
+
+      public static ValueFieldAccessor GetValue(Type probeType) =>
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs:89]: When I did performance testing of this code the non-dictionary version behaved better.
Could you confirm this change is an improvement?

File: Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs (hunk 1 above)
Comment (botinko) [reply Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs:54]:
```
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Running;
using Xtensive.Tuples.Packed;

[DisassemblyDiagnoser(maxDepth: 1)] // change to 0 for just the [Benchmark] method
[MemoryDiagnoser(displayGenColumns: false)]
public class Program
{
  public static void Main(string[] args) =>
    BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args, DefaultConfig.Instance
    );

  [Params(typeof(bool), typeof(short), typeof(TimeSpan), typeof(sbyte?), typeof(ulong?), typeof(TimeSpan?))]
  public Type type;

  [Benchmark]
  public object BenchOld() => TupleLayout.ValueFieldAccessorResolver.GetValue(type);
  
  [Benchmark]
  public object BenchNew() => TupleLayout.ValueFieldAccessorResolver.GetValue2(type);
}
```

For testing

File: Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs (hunk 3 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Tuples/Packed/TupleLayout.cs]: reverted this back to non-dictionary already

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs]: done: `WellKnownTypes.FuncOfTArgTResultType = typeof(Func<,>);`

**Review by AlexUstinov — APPROVED** [review]

**Pull request merged**